### DISCUSSION TITLE
An error occurs when attempting to despawn the entry.

### DISCUSSION THREAD

**Discussion by dolto** [bevyengine/bevy#9701] [Q&A] [answered]
Hello, I am developing BEVY in a web environment.

The following error occurred in the browser console while developing.

![image](https://github.com/bevyengine/bevy/assets/43398206/768620b9-c612-44fe-9331-d9f6e8c3fe56)

After hours of identifying the problem without knowing what the problem was, we found that despawn entity in the web environment resulted in an error.

Below is the code where the error occurred..

```rust
pub fn ui_sidebox_close(
    mut commands: Commands,
    query_sidebox_entity: Query<(Entity, &SideBox)>
    //query: Query<Entity, With<SideBox>> //This also has an error
){
    for (entity, sidebox) in query_sidebox_entity.iter() {
        if sidebox.is_dead {
            if sidebox.size <= 5.001 {
                console::log_1(&JsValue::from_str("try kill the entity..."));
                commands.entity(entity).despawn();
            }
        }
    }

    // for entity in query.iter(){
    //     commands.entity(entity).despawn(); //This also has an error
    // }
}
``` 

As far as I know, there is no problem with the above code in BEVY.

However, I am not very sure because the relevant pages of informal documents are based on 0.9.

How can we solve this problem?
Or if I'm using it incorrectly, I'd appreciate it if you could point it out.

Thank you.

**Comment by tim-blackbird** [comment]
Looks some UI system doesn't handle child entities being despawned here:
https://github.com/bevyengine/bevy/blob/12c6fa7e591545afe8c7101b0f021121eea7272f/crates/bevy_ui/src/update.rs#L40-L41

~~Calling `remove_parent()` before despawning should fix the issue on your end. e.g `commands.entity(entity).remove_parent().despawn();`~~
See rparrett's comment for a better solution.

**Comment by rparrett** [comment, marked as answer]
If `SideBox` exists in a hierarchy, you probably want to use [`despawn_recursive`](https://docs.rs/bevy/latest/bevy/hierarchy/trait.DespawnRecursiveExt.html#tymethod.despawn_recursive).

devil-ira's advice may work as well, but it's likely that your `SideBar` has children that need to be despawned as well. If that's the case, I would use `despawn_recursive`.

The root problem is that `despawn` does not automatically maintain your hierarchy. `despawn_recursive` is "hierarchy aware."

**Reply by dolto** [reply]
When I first designed this UI, I designed it in a way that doesn't remove parents (or children).

However, as the production progressed, there were many things that needed to be fixed, and now the design was changed so that it could be done, but the removal code was not changed...

In the document, BEVY didn't think this would be a problem because he saw that hierarchical relationships between entities only logically look like that, but not in internal systems.

Thank you for your answer!
